Add support for authentication over `core.async_ssh` by password

agoscinski opened the pull request.

agoscinski commented on .github/workflows/ci-code.yml:
```diff
@@ -107,6 +107,9 @@ jobs:
     - name: Setup SSH on localhost
       run: .github/workflows/setup_ssh.sh
 
+    #- name: Setup upterm session
+    #  uses: lhotari/action-upterm@v1
+
```
```suggestion
```

agoscinski commented on src/aiida/orm/authinfos.py:
```diff
@@ -134,10 +135,22 @@ def get_auth_params(self) -> Dict[str, Any]:
+
+        auth_params = copy.deepcopy(auth_params)
+        # the default value for the password in CLI is the empty string
+        # which is covered by evaluating to False in the conditional statement
```
Note that this prevents usage of empty password over CLI. This seems to be a use case of some people looking at the paramiko, but I don't think it is possible with our CLI because you cannot distinguish an empty string and None entry if you want to give prompts for each parameter

unkcpz replied:
This is not clear to me. There are two ways to set answer to CLI prompt without give explicit input, either by press "enter" to continue, which use the default value, or by type `!` to skip the set which I think will set it to `None`. What is the case for the auth_params? I believe for this password which is the password to the remote can never by a empty. So it would make sense to not have default value when press "enter", when user type `!` it should raise.

agoscinski commented:
I am considering to move the password to `AuthInfo` back as it is actually something that is set during configuration and not setup of the computer, and the issues with the CLI should not happen now with the `Password.OBFUSCATED` enum.

unkcpz commented on pyproject.toml:
```diff
@@ -49,7 +49,8 @@ dependencies = [
   'wrapt~=1.11',
-  'chardet~=5.2.0;platform_system=="Windows"'
+  'chardet~=5.2.0;platform_system=="Windows"',
+  'keyringrs~=0.1.0'
```
I release `0.3.1` which include the typing annotation and support for linux if only keyutils can be supported from the platform. Doesn't matter too much if this PR at the moment through.

unkcpz replied:
released https://pypi.org/project/keyringrs/

agoscinski commented:
I talked with @khsrali will the password manager logic to `AuthInfo`, makes more sense. I'll ping reviewers afterwards.

giovannipizzi commented:
Credentials shouldn't be in the computer setup (these are shared between users), but in the AuthInfo (user-specific).
Even better: I do not think we should store credentials in the DB. It should be stored (similar to .ssh keys) in an appropriate file, so that even if the DB is dumped and restored, the credentials are not leaked. Reading data from a DB is much easier than reading a file with proper Unix permissions.

agoscinski commented:
Moved PasswordManager from Computer to AuthInfo. Renamed PasswordManager to SecureStorage. Added more tests to reach higher coverage. Included isinstance check for enums. Added validation step for Linux to check if dbus service is running that is required to use the secure storage on Linux. Discussed with @unkcpz to move to `keyring` to reduce maintenance cost. There is a known memory leak with `keyring`, but it should not be important for us a we do not so many ssh connections within one process. But if people experience issues with this we can move back to `keyring-rs`.

agoscinski commented:
The dbus setup I put into the CI somehow interferes with python xdist or at least with keyring I get segfaults on tests related to the password authentication (but every run different).
With keyringrs I get these errors
```
Platform failure: DBus error: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken
```
```
OSError: Platform failure: DBus error: Failed to activate service 'org.freedesktop.secrets': timed out
```
Seems like the daemon is not reachable

agoscinski commented:
Same error when starting the daemon manually. My setup must be not working properly.

khsrali commented on src/aiida/transports/plugins/ssh_async.py:
```diff
@@ -130,6 +151,21 @@ def __init__(self, *args, **kwargs):
+                )
+            options['password'] = retrieved_password
+        else:
+            # note that this also covers the cases when a not obfuscated password or no password is passed
```
Talked to @agoscinski , we could adopt the comment a little bit
```suggestion
            # note that this also covers the cases when a not obfuscated password or no password is passed
            # also useful, for tests and when plain password is passed via python api (although not supported officially)
```

khsrali commented on tests/transports/test_all_plugins.py:
```diff
@@ -1474,3 +1531,23 @@ async def mock_exec_command_wait_async(*args, **kwargs):
+    if isinstance(custom_transport, LocalTransport):
+        assert 'bash' in goto_computer_cmd
+    else:
+        assert 'ssh' in goto_computer_cmd
```
```suggestion
        assert 'ssh' in goto_computer_cmd
    assert '/path' in goto_computer_cmd
```

khsrali commented on tests/transports/test_all_plugins.py:
```diff
@@ -1474,3 +1531,23 @@ async def mock_exec_command_wait_async(*args, **kwargs):
+        assert 'ssh' in goto_computer_cmd
+
+
+def test_validate_os_supports_secure_storage(monkeypatch):
```
Thank for the test, this only test for failing case. Would be also nice to test a non failing test. 
The actual storage of a password.

khsrali replied:
ok now I see you test that in `tests/orm/test_authinfos.py`
even better!

khsrali commented on tests/transports/test_all_plugins.py:
```diff
@@ -55,30 +57,85 @@ def tmp_path_local(tmp_path_factory):
+    params=[  # first argument specifies transport plugin name the second the, authentication method
+        ('core.local', None),
+        ('core.ssh', None),
+        ('core.ssh_async', 'key'),
+        ('core.ssh_async', 'password-passed'),  # covers the case where the password is passed to transport directly
+        ('core.ssh_async', 'password-from-keychain'),  # covers the case where the password is retrieved from keychain
```
Do we really need to run the entire `test_all_plugins.py` only for various authentication methods?
I mean it's really the connection, only. So one could just check the connection unit (the constructor) individually instead of running for the entire `test_all_plugins.py` I'd say.

unkcpz replied:
I agree with Ali, the `test_all_plugins.py` is quite heavy and not good written. This implementation is only about connection.

khsrali reviewed (changes requested):
Good job, @agoscinski, looks very good!
Please ping me again, once you had a look at that review.
Cheers!

agoscinski commented:
This is the current working theory @unkcpz and me agreed for the cause of errors. It could be still wrong since we did not verified it but it seems to explain the errors quite well, and I would like to move on with this PR.

The current setup of dbus fails using `dbus-x11`
```yml
    sudo apt install -y gnome-keyring dbus-x11
    echo "$(dbus-launch)" >> $GITHUB_ENV
    export $(dbus-launch)
    gnome-keyring-daemon -r -d
    echo 'somecredstorepass' | gnome-keyring-daemon --unlock
```
This is because dbus-x11 is at some point prompting the CI to enter the password after some time. The CI never respons to this prompt and keyring-rs raising this error:
```bash
OSError Platform failure: DBus error: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken
```
The reason why we are using dbus-x11 is because `keyring` cannot directly access the secure storage without a dbus user session. There is always a system dbus session but this one cannot used by `keyring`. In contrast `keyring-rs` can use it to access the secure storage. Therefore the setup in the `keyring-rs` python wrapper is enough to work
```yaml
    sudo apt update -y
    sudo apt install -y libdbus-1-dev libssl-dev gnome-keyring
    gnome-keyring-daemon --components=secrets --daemonize --unlock <<< 'foobar'
```
https://github.com/unkcpz/keyringrs/blob/3e475fcf4844a19eb15c1b1fc48b2987f4c6649d/.github/workflows/CI.yml#L92-L94

This shows that `keyring-rs` is more friendly to be used in a headless mode that is required for the CI. Also it showed that the memory issues in `keyring` are not only related to MacOS. While `keyring-rs` produced proper OSErrors that helped us to figure out the problem, `keyring` just segfaulted. These advantages of `keyring-rs` over `keyring` outweight the additional maintenance costs so we go with `keyring-rs`


Further reading
https://www.baeldung.com/linux/systemd-session-dbus-headless-setup

unkcpz commented on utils/dependency_management.py:
```diff
@@ -171,9 +171,26 @@ def validate_environment_yml():
+                for entry_item in entry_items:
+                    flatten_dependencies.append(entry_item)
+            elif entry.startswith('pip~='):
+                pass
```
```suggestion
                continue
```
I'd avoid `pass` in production code, maybe try to have a more fine control on it? The other part of this change looks good.

unkcpz reviewed (changes requested):
@agoscinski pining me to have a look when it is ready. Great stuff and clean implementation.

khsrali commented:
@agoscinski I'm developing some new features that could benefit from the secure secret storage, that this PR is bringing along. 
May I ask what's your plans on this? I would benefit from this PR moving forward..

agoscinski commented:
superseded by #7492
